File: go/cache/node.go
package cache

import "time"

type Node struct {
	ID          string
	store       *Store
	pubsub      *PubSub
	replicator  *Replicator
	persistence *Persistence
	version     uint64
	clock       func() int64
}

func NewNode(id string, capacity int, clock func() int64, transport Transport, persistence *Persistence) *Node {
	if clock == nil {
		clock = func() int64 { return time.Now().UnixMilli() }
	}
	node := &Node{
		ID:          id,
		store:       NewStore(capacity, nil, clock),
		pubsub:      NewPubSub(),
		replicator:  NewReplicator(transport),
		persistence: persistence,
		clock:       clock,
	}
	if t, ok := transport.(*InMemoryTransport); ok {
		t.Register(id, node.ApplyReplication)
	}
	return node
}

func (n *Node) Load() error {
	if n.persistence == nil {
		return nil
	}
	if err := n.persistence.Load(n.store); err != nil {
		return err
	}
	n.syncVersion()
	return nil
}

func (n *Node) Get(key string) (any, bool) {
	return n.store.Get(key)
}

func (n *Node) Set(key string, value any, ttlMs *int64, replicaIDs []string) uint64 {
	version := n.nextVersion()
	version, evicted := n.store.Set(key, value, ttlMs, &version)
	entry, _ := n.store.GetEntry(key)
	if n.persistence != nil {
		var expiresAt *int64
		if entry.HasExpiry {
			expiresAt = &entry.ExpiresAt
		}
		n.persistence.AppendSet(key, value, expiresAt, version)
	}
	n.replicate(replicaIDs, ReplicationMessage{
		Op:        "set",
		Key:       key,
		Value:     value,
		ExpiresAt: expiresAtFromEntry(entry),
		Version:   version,
	})
	n.recordEvictions(evicted, replicaIDs)
	return version
}

func (n *Node) Delete(key string, replicaIDs []string) bool {
	ok := n.store.Delete(key)
	if !ok {
		return false
	}
	version := n.nextVersion()
	if n.persistence != nil {
		n.persistence.AppendDelete(key, version)
	}
	n.replicate(replicaIDs, ReplicationMessage{Op: "del", Key: key, Version: version})
	return true
}

func (n *Node) Expire(key string, ttlMs int64, replicaIDs []string) bool {
	version := n.nextVersion()
	ok := n.store.Expire(key, ttlMs, &version)
	if !ok {
		return false
	}
	entry, _ := n.store.GetEntry(key)
	if n.persistence != nil && entry.HasExpiry {
		n.persistence.AppendExpire(key, entry.ExpiresAt, version)
	}
	n.replicate(replicaIDs, ReplicationMessage{
		Op:        "expire",
		Key:       key,
		ExpiresAt: expiresAtFromEntry(entry),
		Version:   version,
	})
	return true
}

func (n *Node) Publish(channel string, payload any, replicaIDs []string) int {
	delivered := n.pubsub.Publish(channel, payload)
	n.replicate(replicaIDs, ReplicationMessage{Op: "pub", Channel: channel, Payload: payload})
	return delivered
}

func (n *Node) Subscribe(channel string, handler Subscriber) func() {
	return n.pubsub.Subscribe(channel, handler)
}

func (n *Node) SweepExpired() {
	n.store.SweepExpired()
}

func (n *Node) Snapshot() error {
	if n.persistence == nil {
		return nil
	}
	return n.persistence.Snapshot(n.store)
}

func (n *Node) ApplyReplication(message ReplicationMessage) {
	switch message.Op {
	case "set":
		if message.Key == "" {
			return
		}
		if n.shouldSkip(message.Key, message.Version) {
			return
		}
		ttlMs := (*int64)(nil)
		if message.ExpiresAt != nil {
			value := *message.ExpiresAt - n.clock()
			ttlMs = &value
		}
		version := message.Version
		n.store.Set(message.Key, message.Value, ttlMs, &version)
		n.bumpVersion(message.Version)
	case "del":
		if message.Key == "" {
			return
		}
		if n.shouldSkip(message.Key, message.Version) {
			return
		}
		n.store.Delete(message.Key)
		n.bumpVersion(message.Version)
	case "expire":
		if message.Key == "" || message.ExpiresAt == nil {
			return
		}
		if n.shouldSkip(message.Key, message.Version) {
			return
		}
		ttlMs := *message.ExpiresAt - n.clock()
		version := message.Version
		n.store.Expire(message.Key, ttlMs, &version)
		n.bumpVersion(message.Version)
	case "pub":
		if message.Channel == "" {
			return
		}
		n.pubsub.Publish(message.Channel, message.Payload)
	}
}

func (n *Node) shouldSkip(key string, version uint64) bool {
	if version == 0 {
		return false
	}
	if entry, ok := n.store.GetEntry(key); ok {
		return version <= entry.Version
	}
	return false
}

func (n *Node) replicate(replicaIDs []string, message ReplicationMessage) {
	if len(replicaIDs) == 0 {
		return
	}
	n.replicator.Replicate(replicaIDs, message)
}

func (n *Node) recordEvictions(evicted []string, replicaIDs []string) {
	for _, key := range evicted {
		version := n.nextVersion()
		if n.persistence != nil {
			n.persistence.AppendDelete(key, version)
		}
		n.replicate(replicaIDs, ReplicationMessage{Op: "del", Key: key, Version: version})
	}
}

func (n *Node) nextVersion() uint64 {
	n.version++
	return n.version
}

func (n *Node) bumpVersion(incoming uint64) {
	if incoming > n.version {
		n.version = incoming
	}
}

func (n *Node) syncVersion() {
	var max uint64
	for _, entry := range n.store.Items() {
		if entry.Version > max {
			max = entry.Version
		}
	}
	n.version = max
}

func expiresAtFromEntry(entry Entry) *int64 {
	if entry.HasExpiry {
		return &entry.ExpiresAt
	}
	return nil
}

File: python/cache/node.py
from __future__ import annotations

import time
from typing import Any, Iterable, Optional

from .persistence import Persistence
from .pubsub import PubSub
from .replication import Replicator
from .storage import Store
from .transport import InMemoryTransport, ReplicationMessage, Transport


class Node:
    def __init__(
        self,
        node_id: str,
        *,
        capacity: int = 1024,
        clock: Optional[callable] = None,
        transport: Optional[Transport] = None,
        persistence: Optional[Persistence] = None,
    ) -> None:
        self.id = node_id
        self._clock = clock or (lambda: int(time.time() * 1000))
        self._store = Store(capacity=capacity, clock=self._clock)
        self._pubsub = PubSub()
        self._replicator = Replicator(transport)
        self._persistence = persistence
        self._version = 0
        if isinstance(transport, InMemoryTransport):
            transport.register(self.id, self.apply_replication)

    def load(self) -> None:
        if self._persistence:
            self._persistence.load(self._store)
            self._sync_version()

    def get(self, key: str) -> Optional[Any]:
        return self._store.get(key)

    def set(
        self,
        key: str,
        value: Any,
        ttl_ms: Optional[int] = None,
        *,
        replica_ids: Optional[Iterable[str]] = None,
    ) -> int:
        version = self._next_version()
        version, evicted = self._store.set(key, value, ttl_ms=ttl_ms, version=version)
        entry = self._store.get_entry(key)
        if self._persistence and entry:
            self._persistence.append_set(key, value, entry.expires_at, version)
        self._replicate(
            replica_ids,
            ReplicationMessage(
                op="set",
                key=key,
                value=value,
                expires_at=entry.expires_at if entry else None,
                version=version,
            ),
        )
        self._record_evictions(evicted, replica_ids)
        return version

    def delete(self, key: str, *, replica_ids: Optional[Iterable[str]] = None) -> bool:
        existed = self._store.delete(key)
        if not existed:
            return False
        version = self._next_version()
        if self._persistence:
            self._persistence.append_delete(key, version)
        self._replicate(
            replica_ids,
            ReplicationMessage(op="del", key=key, version=version),
        )
        return True

    def expire(self, key: str, ttl_ms: int, *, replica_ids: Optional[Iterable[str]] = None) -> bool:
        version = self._next_version()
        ok = self._store.expire(key, ttl_ms, version=version)
        if not ok:
            return False
        entry = self._store.get_entry(key)
        if self._persistence and entry and entry.expires_at is not None:
            self._persistence.append_expire(key, entry.expires_at, version)
        self._replicate(
            replica_ids,
            ReplicationMessage(
                op="expire",
                key=key,
                expires_at=entry.expires_at if entry else None,
                version=version,
            ),
        )
        return True

    def publish(
        self,
        channel: str,
        payload: Any,
        *,
        replica_ids: Optional[Iterable[str]] = None,
    ) -> int:
        delivered = self._pubsub.publish(channel, payload)
        self._replicate(
            replica_ids,
            ReplicationMessage(op="pub", channel=channel, payload=payload),
        )
        return delivered

    def subscribe(self, channel: str, handler) -> callable:
        return self._pubsub.subscribe(channel, handler)

    def sweep_expired(self) -> None:
        self._store.sweep_expired()

    def snapshot(self) -> None:
        if self._persistence:
            self._persistence.snapshot(self._store)

    def apply_replication(self, message: ReplicationMessage) -> None:
        if message.op == "set":
            if not message.key:
                return
            if self._should_skip(message.key, message.version):
                return
            ttl_ms = None
            if message.expires_at is not None:
                ttl_ms = max(message.expires_at - self._clock(), 0)
            self._store.set(
                message.key,
                message.value,
                ttl_ms=ttl_ms,
                version=message.version or 0,
            )
            self._bump_version(message.version)
        elif message.op == "del":
            if not message.key:
                return
            if self._should_skip(message.key, message.version):
                return
            self._store.delete(message.key)
            self._bump_version(message.version)
        elif message.op == "expire":
            if not message.key or message.expires_at is None:
                return
            if self._should_skip(message.key, message.version):
                return
            ttl_ms = max(message.expires_at - self._clock(), 0)
            self._store.expire(message.key, ttl_ms, version=message.version or 0)
            self._bump_version(message.version)
        elif message.op == "pub":
            if message.channel is None:
                return
            self._pubsub.publish(message.channel, message.payload)

    def _should_skip(self, key: str, incoming_version: Optional[int]) -> bool:
        if incoming_version is None:
            return False
        existing = self._store.get_entry(key)
        return existing is not None and incoming_version <= existing.version

    def _replicate(
        self, replica_ids: Optional[Iterable[str]], message: ReplicationMessage
    ) -> None:
        if not replica_ids:
            return
        self._replicator.replicate(replica_ids, message)

    def _record_evictions(self, evicted: Iterable[str], replica_ids: Optional[Iterable[str]]) -> None:
        for key in evicted:
            version = self._next_version()
            if self._persistence:
                self._persistence.append_delete(key, version)
            self._replicate(replica_ids, ReplicationMessage(op="del", key=key, version=version))

    def _next_version(self) -> int:
        self._version += 1
        return self._version

    def _bump_version(self, incoming: Optional[int]) -> None:
        if incoming is not None and incoming > self._version:
            self._version = incoming

    def _sync_version(self) -> None:
        max_version = 0
        for _, entry in self._store.items():
            if entry.version > max_version:
                max_version = entry.version
        self._version = max_version
